weave/integrations/openai/openai_sdk.py

```diff
@@ -0,0 +1,275 @@
+
+if typing.TYPE_CHECKING:
+    from openai.types.chat import ChatCompletion, ChatCompletionChunk
+
+
+def openai_on_finish_post_processor(value: "ChatCompletionChunk") -> typing.Dict:
+    from openai.types.chat import ChatCompletion, ChatCompletionChunk
+    from openai.types.chat.chat_completion_chunk import (
+        ChoiceDeltaFunctionCall,
+        ChoiceDeltaToolCall,
+    )
+    from openai.types.chat.chat_completion_message import FunctionCall
+    from openai.types.chat.chat_completion_message_tool_call import (
+        ChatCompletionMessageToolCall,
+        Function,
+    )
+
+    def _get_function_call(
+        function_call: typing.Optional[ChoiceDeltaFunctionCall],
+    ) -> typing.Optional[FunctionCall]:
+        if function_call is None:
+            return function_call
+        if isinstance(function_call, ChoiceDeltaFunctionCall):
+            return FunctionCall(
+                arguments=function_call.arguments,
+                name=function_call.name,
+            )
+        else:
+            return None
+
+    def _get_tool_calls(
+        tool_calls: typing.Optional[typing.List[ChoiceDeltaToolCall]],
+    ) -> typing.Optional[typing.List[ChatCompletionMessageToolCall]]:
+        if tool_calls is None:
+            return tool_calls
+
+        _tool_calls = []
+        if isinstance(tool_calls, list):
+            for tool_call in tool_calls:
+                assert isinstance(tool_call, ChoiceDeltaToolCall)
+                _tool_calls.append(
+                    ChatCompletionMessageToolCall(
+                        id=tool_call.id,
+                        type=tool_call.type,
+                        function=Function(
+                            name=tool_call.function.name,
+                            arguments=tool_call.function.arguments,
+                        ),
+                    )
+                )
+        return _tool_calls
+
+    if isinstance(value, ChatCompletionChunk):
+        final_value = ChatCompletion(
+            id=value.id,
+            choices=[
+                {
+                    "index": choice.index,
+                    "message": {
+                        "content": choice.delta.content,
+                        "role": choice.delta.role,
+                        "function_call": _get_function_call(choice.delta.function_call),
+                        "tool_calls": _get_tool_calls(choice.delta.tool_calls),
+                    },
+                    "logprobs": choice.logprobs,
+                    "finish_reason": choice.finish_reason,
+                }
+                for choice in value.choices
+            ],
+            created=value.created,
+            model=value.model,
+            object="chat.completion",
+            system_fingerprint=value.system_fingerprint,
+            usage=value.usage if hasattr(value, "usage") else None,
+        )
+        return final_value.model_dump(exclude_unset=True, exclude_none=True)
```

Comment:
this dump is unfortunate. I look forward to the day that this is not needed

Reply:
yeah ability to do it server side would be neater I guess.